src/basicfuns.jl

```diff
@@ -11,15 +11,20 @@ julia> StatsFuns.xlogx(0)
 0.0
 ```
 """
-xlogx(x::Real) = x > zero(x) ? x * log(x) : zero(log(x))
+function xlogx(x)
+    result = x * log(x)
+    ifelse(x > zero(x), result, zero(result))
+end
 
 """
     xlogy(x::Real, y::Real)
 
 Return `x * log(y)` for `y > 0` with correct limit at `x = 0`.
 """
-xlogy(x::T, y::T) where {T<:Real} = x > zero(T) ? x * log(y) : zero(log(x))
-xlogy(x::Real, y::Real) = xlogy(promote(x, y)...)
+function xlogy(x, y)
+    result = x * log(y)
+    ifelse(x > zero(x), result, zero(result))
```

Comment:
it does seem a little bit strange to compute the `log` if you're not going to use it.

Reply:
The only case where I don't use it is when `x = 0`, which I think is rare. The upside is that this is branch-free.

Reply:
Wait, is this function even valid for `x < 0`? If not, we should mention that in the docstring.

Reply:
Also this way we get an exception if `y < 0` (even if `x = 0`), which was not the case before. That's a feature I think.

Reply:
Oh yes! Had a lapsus. This should error for `x < 0` and/or `y < 0`. I fixed it.

Reply:
`xlogx(x)` is only defined for `x >= 0`. But `xlogy(x,y)` I think could take negative `x`? There is no reason to impose the artificial constraint `x >= 0`, right ?

Reply:
I just checked that scipy allows negative `x`, https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.xlogy.html

Reply:
I added a test for negative `x` to make sure we keep allowing it in the future.